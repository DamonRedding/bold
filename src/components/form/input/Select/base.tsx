import { createFilter } from 'react-select'
import { Props as ReactSelectProps } from 'react-select/lib/Select'

import { WithStylesProps } from '../../../../styles'
import { Omit } from '../../../../util'
import { InputStatus } from '../TextInput/TextInput'

import { SelectDropdownIndicator, SelectMultiValueRemove, SelectOption } from './components'
import { createSelectStyles } from './createSelectStyle'

export interface DefaultOptionType {
    value: any
    label: string
}

export const defaultSelectFilter = createFilter({
    ignoreAccents: true,
    ignoreCase: true,
})

export const defaultSelectProps: Partial<Omit<ReactSelectProps, 'theme'>> = {
    pageSize: 10,
    backspaceRemovesValue: false,
    isMulti: false,
    isClearable: true,
    placeholder: null,
    loadingMessage: () => 'Carregando...',
    noOptionsMessage: () => 'Nenhum resultado encontrado.',
    getOptionLabel: (option) => option && option.label,
    getOptionValue: (option) => option && option.value,
}

export interface BaseSelectProps<T> extends Omit<ReactSelectProps<T>, 'theme'>, WithStylesProps {
    status?: InputStatus
    disabled?: boolean
}

export function createSelectBaseProps<T>(props: BaseSelectProps<T>): Partial<ReactSelectProps<T>> {
    const { css, theme, status, disabled, components, ...rest } = props
    const styles = createSelectStyles(theme, status === 'error')

    return {
        classNamePrefix: 'react-select',
        styles,
        isDisabled: disabled,
        closeMenuOnSelect: !props.isMulti,
        components: {
            Option: SelectOption,
            DropdownIndicator: SelectDropdownIndicator,
            MultiValueRemove: SelectMultiValueRemove,
            ...components,
        },
        ...rest,
    }
}
