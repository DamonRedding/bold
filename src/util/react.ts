import { MutableRefObject, Ref } from 'react'

export function composeRefs<T>(...refs: Array<Ref<T>>) {
    return (value: T) => refs.forEach(ref => setRef(ref, value))
}

export function setRef<T>(ref: Ref<T>, value: any) {
    if (typeof ref === 'function') {
        ref(value)
    } else if (ref) {
        (ref as MutableRefObject<T>).current = value
    }
}

/**
 * Returns the next sibling of `element` that matches the `predicate`.
 * Returns `null` if not found.
 */
export function getNextSibling(element: Element, predicate: (sibling: Element) => boolean) {
    let current: Element = element

    while (current.nextElementSibling) {
        current = current.nextElementSibling

        if (predicate(current)) {
            return current
        }
    }

    return null
}

/**
 * Returns the previous sibling of `element` that matches the `predicate`.
 * Returns `null` if not found.
 */
export function getPreviousSibling(element: Element, predicate: (sibling: Element) => boolean) {
    let current: Element = element

    while (current.previousElementSibling) {
        current = current.previousElementSibling

        if (predicate(current)) {
            return current
        }
    }

    return null
}
