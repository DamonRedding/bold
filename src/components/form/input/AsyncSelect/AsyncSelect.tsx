import * as React from 'react'
import ReactAsyncSelect, { Props as ReactAsyncSelectProps } from 'react-select/lib/Async'

import { withStyles, WithStylesProps } from '../../../../styles/index'
import { createSelectStyles } from '../Select/createSelectStyle'
import { DefaultOptionType } from '../Select/Select'

export interface AsyncSelectProps<OptionType = DefaultOptionType> extends WithStylesProps,
    ReactAsyncSelectProps<OptionType> {
    status?: '' | 'error'
    disabled?: boolean
}

@withStyles
export class AsyncSelect<OptionType = DefaultOptionType> extends React.Component<AsyncSelectProps<OptionType>> {

    static defaultProps: Partial<AsyncSelectProps<any>> = {
        pageSize: 10,
        backspaceRemovesValue: false,
        isMulti: false,
        isClearable: true,
        placeholder: null,
        loadingMessage: () => 'Carregando...',
        noOptionsMessage: () => 'Nenhum resultado encontrado.',
        getOptionLabel: (option) => option && option.label,
        getOptionValue: (option) => option && option.value,
    }

    render() {
        const { css, theme, status, disabled, ...rest } = this.props

        const styles = createSelectStyles(theme, status === 'error')

        return (
            <ReactAsyncSelect
                classNamePrefix='react-select-async'
                styles={styles}
                isDisabled={disabled}
                closeMenuOnSelect={!this.props.isMulti}
                {...rest}
            />
        )
    }
}
