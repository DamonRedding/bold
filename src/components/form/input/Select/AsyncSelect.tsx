import * as React from 'react'
import ReactAsyncSelect, { Props as ReactAsyncSelectProps } from 'react-select/lib/Async'

import { withStyles, WithStylesProps } from '../../../../styles/index'
import { Omit } from '../../../../util/types'
import { createSelectStyles } from '../Select/createSelectStyle'
import { DefaultOptionType } from '../Select/Select'
import { InputStatus } from '../TextInput/TextInput'

import { SelectDropdownIndicator, SelectMultiValueRemove, SelectOption } from './components'

export interface AsyncSelectProps<OptionType = DefaultOptionType> extends WithStylesProps,
    Omit<ReactAsyncSelectProps<OptionType>, 'theme'> {
    status?: InputStatus
    disabled?: boolean
}

@withStyles
export class AsyncSelect<OptionType = DefaultOptionType> extends React.Component<AsyncSelectProps<OptionType>> {

    static defaultProps: Partial<AsyncSelectProps<any>> = {
        pageSize: 10,
        backspaceRemovesValue: false,
        isMulti: false,
        isClearable: true,
        placeholder: null,
        loadingMessage: () => 'Carregando...',
        noOptionsMessage: () => 'Nenhum resultado encontrado.',
        getOptionLabel: (option) => option && option.label,
        getOptionValue: (option) => option && option.value,
    }

    render() {
        const { css, theme, status, disabled, components, ...rest } = this.props

        const styles = createSelectStyles(theme, status === 'error')

        return (
            <ReactAsyncSelect
                classNamePrefix='react-select-async'
                styles={styles}
                isDisabled={disabled}
                closeMenuOnSelect={!this.props.isMulti}
                components={{
                    Option: SelectOption,
                    DropdownIndicator: SelectDropdownIndicator,
                    MultiValueRemove: SelectMultiValueRemove,
                    ...components,
                }}
                {...rest}
            />
        )
    }
}
