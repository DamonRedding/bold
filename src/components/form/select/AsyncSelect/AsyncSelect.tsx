import * as React from 'react'
import * as Select from 'react-select'

import { withStyles, WithStylesProps } from '../../../../styles/index'

import createSelectStyle from '../createSelectStyle'

export interface AsyncSelectRequestParams {
    query: string
    pageSize: number
}

export interface AsyncSelectProps extends WithStylesProps,
    Pick<Select.ReactAsyncSelectProps,
    'autoload' |
    'backspaceRemoves' |
    'cache' |
    'clearable' |
    'clearValueText' |
    'disabled' |
    'labelKey' |
    'loadingPlaceholder' |
    'ignoreAccents' |
    'ignoreCase' |
    'multi' |
    'noResultsText' |
    'onBlur' |
    'onChange' |
    'placeholder' |
    'searchPromptText' |
    'valueKey'
    > {
    getPage: (params: AsyncSelectRequestParams) => Promise<any>
    maxLength?: number
    pageSize?: number
    searchDelay?: number
    status?: '' | 'error'
    value?: any
}

@withStyles
export class AsyncSelect extends React.Component<AsyncSelectProps> {

    static defaultProps: Partial<AsyncSelectProps> = {
        autoload: false,
        backspaceRemoves: false,
        cache: false,
        clearable: true,
        clearValueText: 'Limpar seleção',
        loadingPlaceholder: 'Carregando...',
        ignoreAccents: false,
        ignoreCase: true,
        multi: false,
        noResultsText: 'Nenhum item encontrado.',
        pageSize: 10,
        placeholder: '',
        searchDelay: 500,
        searchPromptText: 'Digite para pesquisar',
    }

    private typingTimer: number

    render() {
        const { css, theme, status, ...rest } = this.props

        const styles = createSelectStyle(theme)

        const classes = css(styles.default,
            status === 'error' && styles.error)

        return (
            <Select.Async
                className={classes}
                options={[]}
                {...rest}
                inputProps={{ maxLength: this.props.maxLength }}
                loadOptions={this.loadOptions}
                onBlur={this.blur}
            />
        )
    }

    private loadOptions = (query, callback) => {
        if (this.typingTimer) {
            clearTimeout(this.typingTimer)
        }
        if (query !== '') {
            this.typingTimer = window.setTimeout(() => this.getPage(query, callback), this.props.searchDelay)
        } else {
            callback(null, {})
        }
    }

    private getPage = (query, callback) => {
        const params = {
            query,
            pageSize: this.props.pageSize,
        }

        this.props.getPage(params)
            .then(result => {
                const response = {
                    options: result.data,
                }
                callback(null, response)
            }).catch(error => {
                callback(error, null)
            })
    }

    private blur(): React.EventHandler<React.FocusEvent<{}>> {
        return () => {
            if (this.props.onBlur) {
                if (this.props.value) {
                    this.props.onBlur(this.props.value)
                } else {
                    this.props.onBlur(null)
                }
            }
        }
    }
}
