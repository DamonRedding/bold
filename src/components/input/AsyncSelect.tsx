import * as React from 'react'
import * as Select from 'react-select'
import axios from 'axios'

export interface AsyncSelectProps extends Partial<Select.ReactAsyncSelectProps> {
    url: string
    value?: any
    pageSize?: number
    searchDelay?: number
    customizeParams?: any,
    maxLength?: number
}

export class AsyncSelect extends React.Component<AsyncSelectProps> {
    private typingTimer: number

    static defaultProps: Partial<AsyncSelectProps> = {
        autoload: false,
        pageSize: 10,
        searchDelay: 500,
        valueKey: 'id',
        labelKey: 'nome',
        multi: false,
        cache: false,
        placeholder: '',
        clearValueText: 'Limpar seleção',
        loadingPlaceholder: 'Carregando...',
        searchPromptText: 'Digite para pesquisar',
        noResultsText: 'Nenhum item encontrado.',
        backspaceRemoves: false,
        ignoreAccents: false,
        ignoreCase: true,
        clearable: true,
    }

    render() {
        const { value, onBlur, ...rest } = this.props
        return (
            <Select.Async
                options={[]}
                {...rest}
                inputProps={{ maxLength: this.props.maxLength }}
                value={value && value.toJS ? value.toJS() : value}
                onBlur={this.blur()}
                loadOptions={this.loadOptions}
            />
        )
    }

    private loadOptions = (query, callback) => {
        clearTimeout(this.typingTimer)
        if (this.typingTimer) {
            clearTimeout(this.typingTimer)
        }
        this.typingTimer = setTimeout(this.getPage.bind(this, query, callback), this.props.searchDelay)
    }

    private getPage(query, callback) {
        let params = {
            query: query,
            pageSize: this.props.pageSize
        }

        params = this.props.customizeParams ? this.props.customizeParams(params) : params

        return axios.get(this.props.url, { params: params })
            .then(data => {
                let response = {
                    options: data.data
                }
                callback(null, response)
            }).catch(error => {
                callback(error, null)
            })
    }

    private blur = (): React.EventHandler<React.FocusEvent<{}>> => {
        return () => {
            if (this.props.onBlur) {
                if (this.props.value && this.props.value[this.props.labelKey]) {
                    this.props.onBlur(this.props.value)
                } else {
                    this.props.onBlur(null)
                }
            }
        }
    }
}
