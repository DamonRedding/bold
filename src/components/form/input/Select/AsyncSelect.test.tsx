import { mount } from 'enzyme'
import * as React from 'react'

import { withTheme } from '../../../../test'

import { AsyncSelect } from './AsyncSelect'
import { DefaultOptionType } from './base'

interface OptionType extends DefaultOptionType {
    extra: string
}

const options: OptionType[] = [
    { value: 1, label: 'Test #1', extra: 'foo' },
    { value: 2, label: 'Test #2', extra: 'bar' },
]

it('should render correctly', () => {
    const wrapper = mount(withTheme(<AsyncSelect />))
    expect(wrapper.render()).toMatchSnapshot()
})

it('should render correctly when multivalue', () => {
    const wrapper = mount(withTheme(<AsyncSelect<OptionType> options={options} value={[options[0]]} isMulti />))
    expect(wrapper.render()).toMatchSnapshot()
})

it('should accept to override components', () => {
    const wrapper = mount(withTheme(
        <AsyncSelect
            components={{
                Placeholder: (p) => <span className='placeholder' />,
            }}
        />
    ))
    expect(wrapper.render()).toMatchSnapshot()
})
