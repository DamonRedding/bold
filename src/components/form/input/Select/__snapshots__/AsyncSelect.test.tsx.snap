// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should accept to override components 1`] = `
.emotion-6 {
  position: relative;
  box-sizing: border-box;
}

.emotion-5 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: hsl(0,0%,100%);
  border-color: #B9B9BD;
  border-radius: 2px;
  border-style: solid;
  border-width: 1px;
  cursor: text;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  min-height: auto;
  outline: 0 !important;
  position: relative;
  -webkit-transition: all .2s;
  transition: all .2s;
  box-sizing: border-box;
  background: #FFFFFF;
}

.emotion-5:hover {
  border-color: #818489;
}

.emotion-1 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  padding: 2px 8px;
  -webkit-overflow-scrolling: touch;
  position: relative;
  overflow: hidden;
  box-sizing: border-box;
  line-height: 1;
}

.emotion-0 {
  margin: 2px;
  padding-bottom: 2px;
  padding-top: 2px;
  visibility: visible;
  color: hsl(0,0%,20%);
  box-sizing: border-box;
}

.emotion-4 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-align-self: stretch;
  -ms-flex-item-align: stretch;
  align-self: stretch;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  box-sizing: border-box;
}

.emotion-2 {
  -webkit-align-self: stretch;
  -ms-flex-item-align: stretch;
  align-self: stretch;
  background-color: hsl(0,0%,80%);
  margin-bottom: 8px;
  margin-top: 8px;
  width: 1px;
  box-sizing: border-box;
  display: none;
}

.emotion-3 {
  color: #2D343B;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 5px 8px;
  -webkit-transition: color 150ms;
  transition: color 150ms;
  box-sizing: border-box;
  cursor: pointer;
  background: #F2F2F7;
}

.emotion-3:hover {
  color: hsl(0,0%,60%);
}

.emotion-3:hover {
  color: #2D343B;
}

<div
  class="emotion-6"
>
  <div
    class="emotion-5 react-select__control"
  >
    <div
      class="emotion-1 react-select__value-container"
    >
      <span
        class="placeholder"
      />
      <div
        class="emotion-0"
      >
        <div
          class="react-select__input"
          style="display: inline-block;"
        >
          <input
            aria-autocomplete="list"
            autocapitalize="none"
            autocomplete="off"
            autocorrect="off"
            id="react-select-4-input"
            spellcheck="false"
            style="box-sizing: content-box; width: 2px; border: 0px; font-size: inherit; opacity: 1; outline: 0; padding: 0px;"
            tabindex="0"
            type="text"
            value=""
          />
          <div
            style="position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre; font-size: inherit; font-family: \\"IBM Plex Sans\\",sans-serif; letter-spacing: normal; text-transform: none;"
          />
        </div>
      </div>
    </div>
    <div
      class="emotion-4 react-select__indicators"
    >
      <span
        class="emotion-2 react-select__indicator-separator"
      />
      <div
        aria-hidden="true"
        class="emotion-3 react-select__indicator react-select__dropdown-indicator"
      >
        <svg
          height="20"
          viewBox="0 0 24 24"
          width="20"
        >
          <path
            d="M12 15.61l6-7.22H6l6 7.22z"
            data-name="triangle-down"
          />
        </svg>
      </div>
    </div>
  </div>
</div>
`;

exports[`should render correctly 1`] = `
.emotion-7 {
  position: relative;
  box-sizing: border-box;
}

.emotion-6 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: hsl(0,0%,100%);
  border-color: #B9B9BD;
  border-radius: 2px;
  border-style: solid;
  border-width: 1px;
  cursor: text;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  min-height: auto;
  outline: 0 !important;
  position: relative;
  -webkit-transition: all .2s;
  transition: all .2s;
  box-sizing: border-box;
  background: #FFFFFF;
}

.emotion-6:hover {
  border-color: #818489;
}

.emotion-2 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  padding: 2px 8px;
  -webkit-overflow-scrolling: touch;
  position: relative;
  overflow: hidden;
  box-sizing: border-box;
  line-height: 1;
}

.emotion-0 {
  color: hsl(0,0%,50%);
  margin-left: 2px;
  margin-right: 2px;
  position: absolute;
  top: 50%;
  -webkit-transform: translateY(-50%);
  -ms-transform: translateY(-50%);
  transform: translateY(-50%);
  box-sizing: border-box;
}

.emotion-1 {
  margin: 2px;
  padding-bottom: 2px;
  padding-top: 2px;
  visibility: visible;
  color: hsl(0,0%,20%);
  box-sizing: border-box;
}

.emotion-5 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-align-self: stretch;
  -ms-flex-item-align: stretch;
  align-self: stretch;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  box-sizing: border-box;
}

.emotion-3 {
  -webkit-align-self: stretch;
  -ms-flex-item-align: stretch;
  align-self: stretch;
  background-color: hsl(0,0%,80%);
  margin-bottom: 8px;
  margin-top: 8px;
  width: 1px;
  box-sizing: border-box;
  display: none;
}

.emotion-4 {
  color: #2D343B;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 5px 8px;
  -webkit-transition: color 150ms;
  transition: color 150ms;
  box-sizing: border-box;
  cursor: pointer;
  background: #F2F2F7;
}

.emotion-4:hover {
  color: hsl(0,0%,60%);
}

.emotion-4:hover {
  color: #2D343B;
}

<div
  class="emotion-7"
>
  <div
    class="emotion-6 react-select__control"
  >
    <div
      class="emotion-2 react-select__value-container"
    >
      <div
        class="emotion-0 react-select__placeholder"
      />
      <div
        class="emotion-1"
      >
        <div
          class="react-select__input"
          style="display: inline-block;"
        >
          <input
            aria-autocomplete="list"
            autocapitalize="none"
            autocomplete="off"
            autocorrect="off"
            id="react-select-2-input"
            spellcheck="false"
            style="box-sizing: content-box; width: 2px; border: 0px; font-size: inherit; opacity: 1; outline: 0; padding: 0px;"
            tabindex="0"
            type="text"
            value=""
          />
          <div
            style="position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre; font-size: inherit; font-family: \\"IBM Plex Sans\\",sans-serif; letter-spacing: normal; text-transform: none;"
          />
        </div>
      </div>
    </div>
    <div
      class="emotion-5 react-select__indicators"
    >
      <span
        class="emotion-3 react-select__indicator-separator"
      />
      <div
        aria-hidden="true"
        class="emotion-4 react-select__indicator react-select__dropdown-indicator"
      >
        <svg
          height="20"
          viewBox="0 0 24 24"
          width="20"
        >
          <path
            d="M12 15.61l6-7.22H6l6 7.22z"
            data-name="triangle-down"
          />
        </svg>
      </div>
    </div>
  </div>
</div>
`;

exports[`should render correctly when multivalue 1`] = `
.emotion-11 {
  position: relative;
  box-sizing: border-box;
}

.emotion-10 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: hsl(0,0%,100%);
  border-color: #B9B9BD;
  border-radius: 2px;
  border-style: solid;
  border-width: 1px;
  cursor: text;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  min-height: auto;
  outline: 0 !important;
  position: relative;
  -webkit-transition: all .2s;
  transition: all .2s;
  box-sizing: border-box;
  background: #FFFFFF;
}

.emotion-10:hover {
  border-color: #818489;
}

.emotion-3 {
  margin: 2px;
  padding-bottom: 2px;
  padding-top: 2px;
  visibility: visible;
  color: hsl(0,0%,20%);
  box-sizing: border-box;
}

.emotion-9 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-align-self: stretch;
  -ms-flex-item-align: stretch;
  align-self: stretch;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  box-sizing: border-box;
}

.emotion-7 {
  -webkit-align-self: stretch;
  -ms-flex-item-align: stretch;
  align-self: stretch;
  background-color: hsl(0,0%,80%);
  margin-bottom: 8px;
  margin-top: 8px;
  width: 1px;
  box-sizing: border-box;
  display: none;
}

.emotion-8 {
  color: #2D343B;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 5px 8px;
  -webkit-transition: color 150ms;
  transition: color 150ms;
  box-sizing: border-box;
  cursor: pointer;
  background: #F2F2F7;
}

.emotion-8:hover {
  color: hsl(0,0%,60%);
}

.emotion-8:hover {
  color: #2D343B;
}

.emotion-4 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  padding: 2px;
  -webkit-overflow-scrolling: touch;
  position: relative;
  overflow: hidden;
  box-sizing: border-box;
  line-height: 1;
}

.emotion-2 {
  background-color: hsl(0,0%,90%);
  border-radius: 2px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin: 0 2px;
  min-width: 0;
  box-sizing: border-box;
  border: 1px solid #B9B9BD;
}

.emotion-0 {
  border-radius: 2px;
  color: #2D343B;
  font-size: 0.8125rem;
  overflow: hidden;
  padding: 1px 0.5rem;
  padding-left: 6px;
  text-overflow: ellipsis;
  white-space: nowrap;
  box-sizing: border-box;
  background: #FFFFFF;
  border-top-left-radius: 4px;
  border-top-right-radius: 0;
  border-bottom-left-radius: 4px;
  border-bottom-right-radius: 0;
  line-height: 1.25rem;
}

.emotion-1 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-radius: 2px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding-left: 0;
  padding-right: 0;
  box-sizing: border-box;
  cursor: pointer;
  background: #F2F2F7;
  border-top-left-radius: 0;
  border-top-right-radius: 4px;
  border-bottom-left-radius: 0;
  border-bottom-right-radius: 4px;
}

.emotion-1:hover {
  background-color: #FFBDAD;
  color: #DE350B;
}

.emotion-1:hover {
  background: #F2F2F7;
  color: #D83A2E;
}

.emotion-6 {
  color: #B9B9BD;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 5px 8px;
  -webkit-transition: color 150ms;
  transition: color 150ms;
  box-sizing: border-box;
  cursor: pointer;
}

.emotion-6:hover {
  color: hsl(0,0%,60%);
}

.emotion-6:hover {
  color: #D83A2E;
}

.emotion-5 {
  display: inline-block;
  fill: currentColor;
  line-height: 1;
  stroke: currentColor;
  stroke-width: 0;
}

<div
  class="emotion-11"
>
  <div
    class="emotion-10 react-select__control"
  >
    <div
      class="emotion-4 react-select__value-container react-select__value-container--is-multi react-select__value-container--has-value"
    >
      <div
        class="emotion-2 react-select__multi-value"
      >
        <div
          class="emotion-0 react-select__multi-value__label"
        >
          Test #1
        </div>
        <div
          class="emotion-1 react-select__multi-value__remove"
        >
          <svg
            height="22"
            style="fill: currentColor;"
            viewBox="0 0 24 24"
            width="22"
          >
            <path
              d="M18.36 7.05l-1.41-1.41L12 10.59 7.05 5.64 5.64 7.05 10.59 12l-4.95 4.95 1.41 1.41L12 13.41l4.95 4.95 1.41-1.41L13.41 12l4.95-4.95z"
              data-name="times"
            />
          </svg>
        </div>
      </div>
      <div
        class="emotion-3"
      >
        <div
          class="react-select__input"
          style="display: inline-block;"
        >
          <input
            aria-autocomplete="list"
            autocapitalize="none"
            autocomplete="off"
            autocorrect="off"
            id="react-select-3-input"
            spellcheck="false"
            style="box-sizing: content-box; width: 2px; border: 0px; font-size: inherit; opacity: 1; outline: 0; padding: 0px;"
            tabindex="0"
            type="text"
            value=""
          />
          <div
            style="position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre; font-size: inherit; font-family: \\"IBM Plex Sans\\",sans-serif; letter-spacing: normal; text-transform: none;"
          />
        </div>
      </div>
    </div>
    <div
      class="emotion-9 react-select__indicators"
    >
      <div
        aria-hidden="true"
        class="emotion-6 react-select__indicator react-select__clear-indicator"
      >
        <svg
          aria-hidden="true"
          class="emotion-5"
          focusable="false"
          height="20"
          viewBox="0 0 20 20"
          width="20"
        >
          <path
            d="M14.348 14.849c-0.469 0.469-1.229 0.469-1.697 0l-2.651-3.030-2.651 3.029c-0.469 0.469-1.229 0.469-1.697 0-0.469-0.469-0.469-1.229 0-1.697l2.758-3.15-2.759-3.152c-0.469-0.469-0.469-1.228 0-1.697s1.228-0.469 1.697 0l2.652 3.031 2.651-3.031c0.469-0.469 1.228-0.469 1.697 0s0.469 1.229 0 1.697l-2.758 3.152 2.758 3.15c0.469 0.469 0.469 1.229 0 1.698z"
          />
        </svg>
      </div>
      <span
        class="emotion-7 react-select__indicator-separator"
      />
      <div
        aria-hidden="true"
        class="emotion-8 react-select__indicator react-select__dropdown-indicator"
      >
        <svg
          height="20"
          viewBox="0 0 24 24"
          width="20"
        >
          <path
            d="M12 15.61l6-7.22H6l6 7.22z"
            data-name="triangle-down"
          />
        </svg>
      </div>
    </div>
  </div>
</div>
`;
